internal/k8s/controllers/pool_controller_test.go

```diff
@@ -0,0 +1,199 @@
+			desc:                    "handler returns SyncStateSuccess, valid resources",
+			handlerRes:              SyncStateSuccess,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateError, valid resources",
+			handlerRes:              SyncStateError,
+			validResources:          true,
+			expectReconcileFails:    true,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateErrorNoRetry, valid resources",
+			handlerRes:              SyncStateErrorNoRetry,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+		{
+			desc:                    "handler returns SyncStateReprocessAll, valid resources",
+			handlerRes:              SyncStateReprocessAll,
+			validResources:          true,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: true,
+		},
+		{
+			desc:                    "handler returns SyncStateSuccess, invalid resources",
+			handlerRes:              SyncStateSuccess,
+			validResources:          false,
+			expectReconcileFails:    false,
+			expectForceReloadCalled: false,
+		},
+	}
+	for _, test := range tests {
+		var resources metallbcfg.ClusterResources
+		if test.validResources {
+			resources = poolControllerValidResources
+		} else {
+			resources = poolControllerInvalidResources
+		}
+
+		fakeClient := newFakeClient(t, resources)
+
+		expectedCfg, err := config.For(resources, config.DontValidate)
+		if err != nil && test.validResources {
+			t.Fatalf("test %s failed, got unexpected error: %v", test.desc, err)
+		}
+
+		cmpOpt := cmpopts.IgnoreUnexported(metallbcfg.Pool{})
+
+		mockHandler := func(l log.Logger, pools map[string]*config.Pool) SyncState {
+			if !cmp.Equal(expectedCfg.Pools, pools, cmpOpt) {
+				t.Errorf("test %s failed, handler called with unexpected config: %s", test.desc, cmp.Diff(expectedCfg.Pools, pools, cmpOpt))
+			}
+
+			return test.handlerRes
+		}
+
+		calledForceReload := false
+		mockForceReload := func() { calledForceReload = true }
+
+		r := &PoolReconciler{
+			Client:         fakeClient,
+			Logger:         log.NewNopLogger(),
+			Scheme:         scheme,
+			Namespace:      testNamespace,
+			ValidateConfig: config.DontValidate,
+			Handler:        mockHandler,
+			ForceReload:    mockForceReload,
+		}
+		req := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: testNamespace,
+			},
+		}
+
+		_, err = r.Reconcile(context.TODO(), req)
+		failedReconcile := err != nil
+
+		if test.expectReconcileFails != failedReconcile {
+			t.Errorf("%s: fail reconcile expected: %v, got: %v", test.desc, test.expectReconcileFails, failedReconcile)
+		}
+
+		if test.expectForceReloadCalled != calledForceReload {
+			t.Errorf("%s: call force reload expected: %v, got: %v", test.desc, test.expectForceReloadCalled, calledForceReload)
+		}
+	}
+}
+
+var (
```

Comment:
should we keep the vars at the top of the file like we usually do?

Reply:
I suggested doing it this way. We refer to the variables by their name, and makes sense to understand what the test does when we open the file, so we can see at glance what the tests are.